Rewrite inventory/ini.py

amenonsen opened the pull request:
The new code parses INI-format inventory files in a single pass using a well-documented state machine that reports precise errors and eliminates the duplications and inconsistencies and outright errors in the earlier three-phase parsing code (e.g. three ways to skip comments). It is also much easier now to follow what decisions are being taken on the basis of the parsed data.

This has a :+1: from @srvg after review and testing.

cc: @abadger @bcoca @jimi-c

amenonsen committed:
Rewrite the INI InventoryParser

The new code parses INI-format inventory files in a single pass using a
well-documented state machine that reports precise errors and eliminates
the duplications and inconsistencies and outright errors in the earlier
three-phase parsing code (e.g. three ways to skip comments). It is also
much easier now to follow what decisions are being taken on the basis of
the parsed data. The comments point out various potential improvements,
particularly in the area of consistent IPv6 handling.

On the ornate marble tombstone of the old code, the following
inscription is one last baffling memento from a bygone age:

-    def _before_comment(self, msg):
-        ''' what's the part of a string before a comment? '''
-        msg = msg.replace("\#","**NOT_A_COMMENT**")
-        msg = msg.split("#")[0]
-        msg = msg.replace("**NOT_A_COMMENT**","#")
-        return msg

amenonsen committed:
Rename 'section' to 'groupname' to better reflect its purpose

amenonsen committed:
Make _parse take an array of input lines as an argument

(There's no compelling reason to do this right now, but should be parser
need to be called multiple times in future, this makes it easier.)

amenonsen committed:
Use a self._raise_error helper and avoid passing the lineno around

Based on a patch by @Richard2ndQuadrant.

srgvg commented:
:+1:  tested it by dumping my big inventory with lots of ini's into a json, and compared versions with old and new ini.py, which are identical

amenonsen committed:
Sanitize IPv6 hostname/port handling

Now we accept IPv6 addresses _with port numbers_ only in the standard
[xxx]:NN notation (though bare IPv6 addresses may be given, as before,
and non-IPv6 addresses may also be placed in square brackets), and any
other host identifiers (IPv4/hostname/host pattern) as before, with an
optional :NN suffix.

abadger commented:
Thanks for the code cleanup, fixes, and testing guys!

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
